File: src/nc150/cppsols/mergeKLists.cpp
#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
using namespace std;
using namespace __gnu_pbds;
template<class T> using ordered_set = tree<T, null_type, less<T>, rb_tree_tag, tree_order_statistics_node_update>;
#define M_PI 3.14159265358979323846
#define MOD 1000000007
#define INF 1000000005
#define NEG_INF -1000000005
#define sz(x) (int)x.size()
#define all(x) x.begin(), x.end()
typedef long long ll;
typedef vector<int> vint;
typedef vector<vint> vvint;
typedef vector<ll> vll;
typedef vector<vll> vvll;
typedef pair<int, int> ipair;
typedef pair<ll, ll> llpair;

void __print(int x) {cerr << x;}
void __print(long x) {cerr << x;}
void __print(long long x) {cerr << x;}
void __print(unsigned x) {cerr << x;}
void __print(unsigned long x) {cerr << x;}
void __print(unsigned long long x) {cerr << x;}
void __print(float x) {cerr << x;}
void __print(double x) {cerr << x;}
void __print(long double x) {cerr << x;}
void __print(char x) {cerr << '\'' << x << '\'';}
void __print(const char *x) {cerr << '\"' << x << '\"';}
void __print(const string &x) {cerr << '\"' << x << '\"';}
void __print(bool x) {cerr << (x ? "true" : "false");}

template<typename T, typename V>
void __print(const pair<T, V> &x) {cerr << '{'; __print(x.first); cerr << ','; __print(x.second); cerr << '}';}
template<typename T>
void __print(const T &x) {int f = 0; cerr << '{'; for (auto &i: x) cerr << (f++ ? "," : ""), __print(i); cerr << "}";}
void _print() {cerr << "]\n";}
template <typename T, typename... V>
void _print(T t, V... v) {__print(t); if (sizeof...(v)) cerr << ", "; _print(v...);}
#ifndef ONLINE_JUDGE
#define debug(x...) cerr << "[" << #x << "] = ["; _print(x)
#else
#define debug(x...)
#endif

/*
	Author: Koushik Sahu
	Created: 12:31:53 AM(00:31:53) IST(+05:30) 24-10-2025 Fri
*/

struct ListNode {
	int val;
	ListNode *next;
	ListNode() : val(0), next(nullptr) {}
	ListNode(int x) : val(x), next(nullptr) {}
	ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Comparator {
public:
	bool operator()(ListNode* a, ListNode* b) {
		return a->val > b->val;
	}
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
		priority_queue<ListNode*, vector<ListNode*>, Comparator> pq;
		for(auto list: lists){
			if(list != nullptr){
			    pq.push(list);
            }
		}
		ListNode* ans = nullptr;
		ListNode* lst = nullptr;
		while(!pq.empty()){
			ListNode* curr = pq.top();
			pq.pop();
			if(lst == nullptr) {
				ans = new ListNode(curr->val, nullptr);
				lst = ans;
			} else {
				lst->next = new ListNode(curr->val, nullptr);
				lst = lst->next;
			}
			ListNode* tmp = curr;
			curr = curr->next;
			delete tmp;
			if(curr != nullptr) {
				pq.push(curr);
			}
		}
		return ans;
    }
};

int main(){
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	Solution s;
	vector<ListNode*> a;
	auto ans = s.mergeKLists(a);
	return 0;
}

File: src/nc150/cmd/mergeKLists.go
package main

func mergeKLists(lists []*ListNode) *ListNode {
	if len(lists) == 1 {
		return lists[0]
	}
	n := len(lists)
	list1 := lists[0]
	list2 := mergeKLists(lists[1:n])
	ans := (*ListNode)(nil)
	lst := (*ListNode)(nil)
	for list1 != nil && list2 != nil {
		if list1.Val < list2.Val {
			if ans == nil {
				ans = &ListNode {Val: list1.Val, Next: nil}
				lst = ans
			} else {
				lst.Next = &ListNode {Val: list1.Val, Next: nil}
				lst = lst.Next
			}
			list1 = list1.Next
		} else {
			if ans == nil {
				ans = &ListNode {Val: list2.Val, Next: nil}
				lst = ans
			} else {
				lst.Next = &ListNode {Val: list2.Val, Next: nil}
				lst = lst.Next
			}
			list2 = list2.Next
		}
	}
	for list1 != nil {
		if ans == nil {
			ans = &ListNode {Val: list1.Val, Next: nil}
			lst = ans
		} else {
			lst.Next = &ListNode {Val: list1.Val, Next: nil}
			lst = lst.Next
		}
		list1 = list1.Next
	}
	for list2 != nil {
		if ans == nil {
			ans = &ListNode {Val: list2.Val, Next: nil}
			lst = ans
		} else {
			lst.Next = &ListNode {Val: list2.Val, Next: nil}
			lst = lst.Next
		}
		list2 = list2.Next
	}
	return ans
}
